syntax/go.vim

```diff
@@ -162,15 +162,17 @@ endif
 syn match       goSingleDecl        /\%(import\|var\|const\) [^(]\@=/ contains=goImport,goVar,goConst
 
 " Integers
-syn match       goDecimalInt        "\<-\=\(0\|[1-9]\d*\)\%([Ee][-+]\=\d\+\)\=\>"
-syn match       goHexadecimalInt    "\<-\=0[xX]\x\+\>"
-syn match       goHexadecimalError  "\<-\=0[xX]\x*[^ \t0-9A-Fa-f]\S*\>"
-syn match       goOctalInt          "\<-\=0\o\+\>"
-syn match       goOctalError        "\<-\=0[XxBb\]]\@!\o*[^ \t0-7]\S*\>"
-syn match       goBinaryInt         "\<-\=0[bB][01]\+\>"
-syn match       goBinaryError       "\<-\=0[bB][01]*[^ \t01]\S*\>"
+syn match       goDecimalInt        "\<-\=\(0\|[1-9]_\?\(\d\+_\?\)*\)\%([Ee][-+]\=\d\+\)\=\>"
+syn match       goDecimalErrorInt   "\<-\=\(_\(\d\+_*\)\+\|\(\d\+_*\)\+_\{2,\}\(\d\+_*\)\+\|\(\d\+_*\)\+_\+\)\%([Ee][-+]\=\d\+\)\=\>"
+syn match       goHexadecimalInt    "\<-\=0[xX]_\?\(\x\+_\?\)\+\>"
+syn match       goHexadecimalError  "\<-\=0[xX]_\?\(\x\+_\?\)*\(\([^ \t0-9A-Fa-f_]\|_\{2,\}\)\S*\|_\)\>"
+syn match       goOctalInt          "\<-\=0[oO]\?_\?\(\o\+_\?\)\+\>"
+syn match       goOctalError        "\<-\=0\([0-7oO_]*\([^ \t0-7oOxX\]\}\_]\+\|[oO]\{2,\}\|_\{2,\}\)[0-7oO_]*\)\+\S*\>"
```

Comment:
Why does this have `the xX` in `[^ \t0-7oOxX\]`? And why is there a `\}` immediately following that set?

Why is this checking for double prefixes?

Why doesn't this have the same kind of check for trailing underscores that goHexadecimalError and goBinaryError have?

Reply:
We need the `xX` to doesn't match on valid hexadecimal smth like: `0x123`.

I change that regexp a bit so it doesn't have `\}` anymore. I don't have much experience with vimsrcipt just had tried to exclude something like `[]int{0}` but understood that it doesn't match anyway.

> Why is this checking for double prefixes?

Cuz it is forbidden to use double prefixes. e.g. `0oo123` syntaxly incorrect.

>Why doesn't this have the same kind of check for trailing underscores that goHexadecimalError and goBinaryError have?

Thanks, that's was fixed.

Reply:
> We need the xX to doesn't match on valid hexadecimal smth like: 0x123

Then shouldn't it also have `Bb` to avoid matching binary literals?

> Cuz it is forbidden to use double prefixes. e.g. 0oo123 syntaxly incorrect.

I understand that. My question is really why a check for the double prefix is needed given that it `0oo123` is an error just as `0oa123` is an error

Reply:
You're right that `[]int{0} ` needs to be handled correctly, as does references to a zeroth index in a slice.

I'll add tests for each of those.

Reply:
1. If we apart this regexp than yes, indeed. But I believed that we also account for following overrides. Otherwise, we might need to fix up a decimal and binary error as well to do not match on some floats and imaginary.
2. The previous or-brach (`[^ \t0-7oOxX_/]`) matches on `0oa123` but not on `0oo123`, so we need to check that case specifically.